skimage/morphology/_skeletonize_3d_cy.pyx.in

```diff
@@ -0,0 +1,434 @@
+    # because of `simple_border_points` being a python list which is being
+    # mutated.
+    cdef:
+        list simple_border_points = []
+        pixel_type neighborhood[27]
+        npy_intp p, r, c
+        bint is_border_pt
+        (npy_intp, npy_intp, npy_intp) point
+
+        # rebind a global name to avoid lookup. The table is filled in
+        # at import time.
+        int[::1] Euler_LUT = LUT
+
+    # loop through the image
+    # NB: each loop is from 1 to size-1: img is padded from all sides 
+    for p in range(1, img.shape[0] - 1):
+        for r in range(1, img.shape[1] - 1):
+            for c in range(1, img.shape[2] - 1):
+
+                # check if pixel is foreground
+                if img[p, r, c] != 1:
+                    continue
+
+                is_border_pt = (curr_border == 1 and img[p, r, c-1] <= 0 or  #N
+                                curr_border == 2 and img[p, r, c+1] <= 0 or  #S
+                                curr_border == 3 and img[p, r+1, c] <= 0 or  #E
+                                curr_border == 4 and img[p, r-1, c] <= 0 or  #W
+                                curr_border == 5 and img[p+1, r, c] <= 0 or  #U
+                                curr_border == 6 and img[p-1, r, c] <= 0)    #B
+                if not is_border_pt:
+                    # current point is not deletable
+                    continue
+
+                get_neighborhood(img, p, r, c, neighborhood)
+
+                # check if (p, r, c) is an endpoint (then it's not deletable.)
+                if is_endpoint(neighborhood):
+                    continue
+
+                # check if point is Euler invariant (condition 1 in [Lee94]_):
+                # if it is not, it's not deletable.
+                if not is_Euler_invariant(neighborhood, Euler_LUT):
+                    continue
+
+                # check if point is simple (i.e., deletion does not
+                # change connectivity in the 3x3x3 neighborhood)
+                # this are conditions 2 and 3 in [Lee94]_
+                if not is_simple_point(neighborhood):
+                    continue
+
+                # ok, add (p, r, c) to the list of simple border points
+                point = p, r, c
+                simple_border_points.append(point)
+    return simple_border_points
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef void get_neighborhood(pixel_type[:, :, ::1] img,
+                           npy_intp p, npy_intp r, npy_intp c,
+                           pixel_type neighborhood[]):
+    """Get the neighborhood of a pixel.
+
+    Assume zero boundary conditions. 
+    Image is already padded, so no out-of-bounds checking.
+
+    For the numbering of points see Fig. 1a. of [Lee94]_, where the numbers
+    do *not* include the center point itself. OTOH, this numbering below
+    includes it as number 13. The latter is consistent with [IAC15]_.
+    """
+    neighborhood[0] = img[p-1, r-1, c-1]
```

Comment:
Agreed, it would be nice indeed. Several reasons why it's not done though: first, the indices in the rest of the code depend on this precise numbering, so being explicit here might be good from pure maintenance POV. Second, using memoryviews (let alone numpy arrays) is creating major perf bottlenecks. So there's a benefit to just using a plain C array.